server.js

```diff
@@ -1,27 +1,184 @@
+    type: Number, // It must be a number
+    default: 0 // The default value is 0
+  },
+  name: { // This will store the author's name
+    type: String, // It must be a string
+    required: true, // It's required
+    minlength: 2, // Minimum length is 2 characters
+    maxlength: 14 // Maximum length is 14 characters
+  },
+  createdAt: { // This will store the date of the thought creation
+    type: Date, // It's a date
+    default: Date.now // The default value is the current date
+  }
+});
+
+// Creating a mongoose model from our schema
+const Thought = mongoose.model("Thought", thoughtSchema);
+
+// The port our server will listen to. Defaults to 8080 if PORT environment variable is not set
 const port = process.env.PORT || 8080;
+
+// Creating our express server
 const app = express();
 
-// Add middlewares to enable cors and json body parsing
+// Using cors middleware to handle cross-origin requests
 app.use(cors());
+// Using express's json middleware to parse JSON bodies of incoming requests
 app.use(express.json());
 
-// Start defining your routes here
+// We're using the 'express-list-endpoints' package to list all endpoints on the root route ("/")
+const listEndpoints = require('express-list-endpoints');
+
+// Defining the root route. It responds with a list of all routes
 app.get("/", (req, res) => {
-  res.send("Hello Technigo!");
+  res.send(listEndpoints(app));
+});
+
+// Defining the "/thoughts" GET route. It responds with a list of 20 thoughts, sorted by creation date
+app.get("/thoughts", async (req, res) => {
+  const thoughts = await Thought.find().sort({ createdAt: 'desc' }).limit(20).exec();
+  res.json(thoughts);
+});
+
+// Defining the "/thoughts" POST route. It saves a new thought and responds with the saved thought
+app.post('/thoughts', async (req, res) => {
+  // Retreieve the information sent by the client to our API endpoint:
+  const { message, name } = req.body;
+  // Use our mongoose model to create the database entry:
+  const thought = new Thought({ message, name });
+  try {
+    // Save the thought to the database
+    const savedThought = await thought.save();
+    // Respond with the saved thought
+    res.status(201).json(savedThought);
+  } catch (err) {
+    // If there was an error, respond with the error message and a 400 status code
+    res.status(400).json({ message: 'Could not save thought', errors: err.errors });
+  }
+});
+
+// Defining the "/thoughts/:_id" GET route. It responds with a single thought based on the id
+app.get('/thoughts/:_id', async (req, res) => {
+  try {
+    const singleThought = await Thought.findById(req.params._id);
+    if (singleThought) {
+      res.status(200).json({
+        message: "One single thought",
+        success: true,
+        body: singleThought
+      });
+    } else {
+      res.status(404).json({
+        success: false,
+        body: {
+          message: "Couldn't find one single thought. Double check the ID you are using"
```

Comment:
This is also a good message for the user! So that they know what they're doing wrong.